subprojects/docs/src/docs/userguide/jvm/java_library_plugin.adoc

```diff
@@ -241,6 +241,146 @@ The role of each configuration is described in the following tables:
 [[sec:java_library_modular]]
 == Building Modules for the Java Module System
 
+Samples: link:../samples/#java_modules[Java Modules sample projects]
+
+[NOTE]
+====
+Java Module System support is an incubating feature and therefore you need to turn on _module path inference_ explicitly as shown below.
+====
+
+.Activate module path inference
+====
+include::sample[dir="snippets/java-library/module/groovy",files="build.gradle[tags=inferModulePath]"]
+include::sample[dir="snippets/java-library/module/kotlin",files="build.gradle.kts[tags=inferModulePath]"]
+====
+
+Since Java 9, Java itself offers a https://www.oracle.com/corporate/features/understanding-java-9-modules.html[module system] that allows for strict encapsulation during compile and runtime.
+You can turn a Java library into a _Java Module_ by creating a `module-info.java` file in the `main/java` source folder.
+
+```
+src
+└── main
+    └── java
+        └── module-info.java
+```
+
+In the module info file, you declare a _module name_, which packages of your module you want to _export_ and which other modules you _require_.
+
+[source,java,indent=0]
+.module-info.java file
+----
+include::{samplesPath}/java-library/module/groovy/src/main/java/module-info.java[]
+----
+
+To tell the Java compiler that a Jar is a module, as opposed to a traditional Java library, Gradle needs to place it on the so called _module path_.
+It is an alternative to the _classpath_, which is the traditional way to tell the compiler about compiled dependencies.
+Gradle will automatically put a Jar of your dependencies on the module path, instead of the classpath, if:
+
+* module path inference is turned on via `java. modularity.inferModulePath.set(true)` (Java Module support is currently an incubating feature; in future Gradle versions this will become the default)
```

Comment:
Okay it was not clear then. :) There are **three** things that need to be true

- Turned on
- The project **you build** needs to be a module (otherwise everything is classpath as before)
- The Jar we look at here needs to be a module.

I'll see how I can reword this.

Reply:
Ah right, my bad.

Still, I believe that the opt-in should be separated from the list since that's temporary.

Reply:
I think I'd like to keep the list to make it clear for users who want to understand what's going on here behind the scenes. I made it more explicit that all three things need to be true and explained each point clearer (or so I hope).

Reply:
>  I believe that the opt-in should be separated from the list since that's temporary.

I would keep it in like this for now. I think it's easy to forget and stumble over it. We can remove it with 7.0: https://github.com/gradle/gradle/issues/12742